### PR TITLE
MWC: make Dummy function placeholders callable to fix build with MSVC

### PR DESCRIPTION
**Pull request by 678098**
Simpler to observe what's going on in this source: https://github.com/bloomberg/blazingmq/blob/49fe8ab987bdaeef1298bb16538b373922b8d73f/src/groups/mwc/mwcex/mwcex_job.h

Note that in the code we instantiate some template classes with `Dummy` objects, like this:
```
mwcu::ObjectPlaceHolder<sizeof(Job_Target<Dummy>)> d_target;
```

We don't actually use these `Job_Target<Dummy>` objects, but let's look closely what do we want to instantiate:

```
bslmf::Util::moveIfSupported(d_function.object())();
```
Here:
https://github.com/bloomberg/blazingmq/blob/49fe8ab987bdaeef1298bb16538b373922b8d73f/src/groups/mwc/mwcex/mwcex_job.h#L217-L222

So basically in this code we are trying to "call" a `Dummy` structure here: bslmf::Util::moveIfSupported(d_function.object())**()**;

I am curious why it even compiled before.

Note that we already have this fix in some, but not all the places where it's needed. For example, here: https://github.com/bloomberg/blazingmq/blob/7f817c436d2e444969ad6e673032aaa9a09412b5/src/groups/mwc/mwcu/mwcu_operationchain.h#L390-L392

**UPD:** for executor it's even more complicated, since we have to implement the common interface of `Executor_TargetBase` which is expected as a template parameter.

### CODE REVIEW

**Comment by hallfox** [comment]
Why was MSVC fixed by making these types callable? I'm a little concerned we're taking a PoD type and unmaking it PoD.

**Comment by 678098** [comment]
@hallfox added an explanation for this to the title of the PR

**Comment by ddragan-bloomberg** [comment]
I'll review this.

**Comment by ddragan-bloomberg** [comment]
Template code is not instantiated unless used explicitly, which I believe is the reason this code compiles perfectly in GCC/Clang. MSVC might by buggy. Does `libmwc` even have to compile on Windows? If so, what is the minimum compiler version is has to compile with?

File: src/groups/mwc/mwcex/mwcex_future.h [hunk 1]
```diff
@@ -232,7 +232,7 @@ class Future_Exception {
 
     /// Provides a "small" dummy object which size is used to calculate the
     /// size of the on-stack buffer used for optimization.
-    struct Dummy {
+    struct Dummy : public mwcu::NoOp {
```
Comment (ddragan-bloomberg) [inline src/groups/mwc/mwcex/mwcex_future.h:235]: I kinda get the rational behind changes in `mwcex_job` -- the `Dummy` is involved with a functor, but what is the rational here? Does the code not compile if you leave this dummy (and the one below) as is?

**Comment by ddragan-bloomberg** [comment]
Side note. Evgenii, while you're at it, might as well replace usage of `mwcu::NoOp` with `bdlf::NoOp`

File: .gitignore [hunk 2]
```diff
@@ -26,3 +26,15 @@ settings.json
 .pytest_cache/
 .cache/
 .venv/
+
+# Symlink from 'src/applications/bmqbrkr/run'
+src/applications/bmqbrkr/etc/etc
+
+# 'sim_cpp11_features.pl' backups
+src/groups/mwc/mwcex/mwcex_bindutil.h.bak
```
Comment (ddragan-bloomberg) [inline .gitignore:34]: wouldn't it be simpler to remove those backup files in that same script that calls sim_cpp11_features.pl ?

**Comment by ddragan-bloomberg** [comment]
Evgenii, since you decided to replace `mwcu::NoOp` in all `mwc` and `BMQ` code, please do this in a separate PR, then rebase this PR on top of the other one once I merge it.

And DO NOT remove the `mwcu_noop` component! Other libraries and apps depend on `libmwc` and might be still using `mwcu::NoOp`.

File: .gitignore (hunk 2 above)
Comment (678098) [reply .gitignore]: We don't have access to the script that launches the `sim_cpp11_features.pl`, since it's done within `bde-tools`. I updated `.gitignore` with wildcard backup pattern which covers all the changes, and this is the simplest way.

**Comment by 678098** [comment]
> And DO NOT remove the `mwcu_noop` component! Other libraries and apps depend on `libmwc` and might be still using `mwcu::NoOp`.

I already checked that no other libraries depend on `mwcu::NoOp`

**Comment by ddragan-bloomberg** [comment]
> I already checked that no other libraries depend on mwcu::NoOp

Okay, It that case we are good to remove it.

**Comment by ddragan-bloomberg** [comment]
Also note than, unlike `mwc`, BDE provides s static instance of `bdlf::NoOp` -- `bdlf::noOp`, should use it where appropriate.

**Comment by 678098** [comment]
> Also note than, unlike `mwc`, BDE provides s static constexpr instance of `bdlf::NoOp` -- `bdlf::noOp`, should use it where appropriate.

This is a good point. We might save a few ticks in some places.

> Evgenii, since you decided to replace `mwcu::NoOp` in all `mwc` and `BMQ` code, please do this in a separate PR, then rebase this PR on top of the other one once I merge it.

Moved `mwcu::NoOp` -> `bdlf::NoOp` transition from this branch.

> Does `libmwc` even have to compile on Windows?

We can get Windows support for `mwc` almost for free, a bit more complicate for lib `bmq` and even more complicated for `mqb`. So, we can pretty easily make buildable Windows SDK (depends on `mwc` + `bmq`), and it's more difficult to build a Windows Broker (depends on `mwc` + `bmq` + `mqb`).

For `mwc`, just fix these few templates and that's it. We don't have requirement to build BlazingMQ on Windows, but we can at least gradually extend support for this platform, by fixing a few non-portable parts of the codebase. There are external users who might benefit from using it on Windows https://github.com/bloomberg/blazingmq/issues/314

> If so, what is the minimum compiler version is has to compile with?

No strict requirements. On this stage, it's worth it to solve problems once they occur, until we have a better vision.

**Comment by ddragan-bloomberg** [comment]
How do you check the code compiles on Windows. Which compiler do you use? Can you use a newer compiler?

**Comment by 678098** [comment]
> How do you check the code compiles on Windows. Which compiler do you use? Can you use a newer compiler?

I have a configured Windows preset which I can build. Also, I use the latest MSVC compiler available.

```
$ "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\bin\HostX64\x64\CL.exe"
Microsoft (R) C/C++ Optimizing Compiler Version 19.40.33812 for x64
```

If I roll back the proposed changes, I will get errors from the compiler:

```
D:\blazingmq\src\groups\mwc\mwcex\mwcex_job.h(224,33): error C2064: term does not evaluate to a function taking 0 arguments [D:\blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
  mwcex_promise.cpp
  mwcex_promise_cpp03.cpp
  mwcex_sequentialcontext.cpp
D:\blazingmq\src\groups\mwc\mwcex\mwcex_job.h(224,33): error C2064: term does not evaluate to a function taking 0 arguments [D:\blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
  mwcex_strand.cpp
D:\blazingmq\src\groups\mwc\mwcex\mwcex_job.h(224,33): error C2064: term does not evaluate to a function taking 0 arguments [D:\blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
  mwcex_systemexecutor.cpp
  Generating Code...
```

**Comment by 678098** [comment]
> How do you check the code compiles on Windows. Which compiler do you use? Can you use a newer compiler?

I have a configured Windows preset which I can build. Also, I use the latest MSVC compiler available:

```
$ "Tools\MSVC\14.40.33807\bin\HostX64\x64\CL.exe"
Microsoft (R) C/C++ Optimizing Compiler Version 19.40.33812 for x64
```

If I roll back the proposed changes, I will get errors from the compiler:

```
D:\blazingmq\src\groups\mwc\mwcex\mwcex_job.h(224,33): error C2064: term does not evaluate to a function taking 0 arguments [D:\blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
  mwcex_promise.cpp
  mwcex_promise_cpp03.cpp
  mwcex_sequentialcontext.cpp
D:\blazingmq\src\groups\mwc\mwcex\mwcex_job.h(224,33): error C2064: term does not evaluate to a function taking 0 arguments [D:\blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
  mwcex_strand.cpp
D:\blazingmq\src\groups\mwc\mwcex\mwcex_job.h(224,33): error C2064: term does not evaluate to a function taking 0 arguments [D:\blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
  mwcex_systemexecutor.cpp
  Generating Code...
```

**Comment by ddragan-bloomberg** [comment]
There is a question left unanswered -- what is the rational behind changes in `mwcex_future.h`? Does the code compiles without them?

**Comment by 678098** [comment]
> There is a question left unanswered -- what is the rational behind changes in `mwcex_future.h`? Does the code compiles without them?

It doesn't

**Comment by ddragan-bloomberg** [comment]
I assume this is because `Future_Callback` is functor-related. What about `Future_Exception`? Are you able to remove the use of `NoOp` there?

**Comment by 678098** [comment]
> I assume this is because `Future_Callback` is functor-related. What about `Future_Exception`? Are you able to remove the use of `NoOp` there?

No

If I remove inheritance from `NoOp` in either `Future_Callback` or `Future_Exception`, it will cause the same compilation error in `Job_Target::invoke()`:

```
// MANIPULATORS
template <class FUNCTION>
inline void Job_Target<FUNCTION>::invoke()
{
    bslmf::Util::moveIfSupported(d_function.object())(); // blazingmq\src\groups\mwc\mwcex\mwcex_job.h(224,33): error C2064: term does not evaluate to a function taking 0 arguments [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]

}
```

And regarding the `Executor_Box_SboImp::Dummy`, if left as-is, the compilation errors would be:

```
blazingmq\src\groups\mwc\mwcex\mwcex_executor.h(603,52): error C2676: binary '==': 'const EXECUTOR' does not define this operator or a conversion to a type acceptable to the predefined operator [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
blazingmq\src\groups\mwc\mwcex\mwcex_executor.h(603,52): error C2676:         with [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
blazingmq\src\groups\mwc\mwcex\mwcex_executor.h(603,52): error C2676:         [ [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
blazingmq\src\groups\mwc\mwcex\mwcex_executor.h(603,52): error C2676:             EXECUTOR=BloombergLP::mwcex::Executor_Box_SboImp::Dummy [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
blazingmq\src\groups\mwc\mwcex\mwcex_executor.h(603,52): error C2676:         ] [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
```

I found out that it's not necessary to inherit `Executor_Box_SboImp::Dummy` from `Executor_TargetBase`, but I had to provide implementations of two methods which causes compilation errors:

```
blazingmq\src\groups\mwc\mwcex\mwcex_executor.h(639,52): error C2676: binary '==': 'const EXECUTOR' does not define this operator or a conversion to a type acceptable to the predefined operator [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]

blazingmq\src\groups\mwc\mwcex\mwcex_executortraits.h(206,14): error C2039: 'post': is not a member of 'BloombergLP::mwcex::Executor_Box_SboImp::Dummy' [blazingmq\cmake.bld\src\groups\mwc\mwcex-iface.vcxproj]
```

File: .gitignore [hunk 3]
```diff
@@ -26,3 +26,10 @@ settings.json
 .pytest_cache/
 .cache/
 .venv/
+
+# Symlink from 'src/applications/bmqbrkr/run'
+src/applications/bmqbrkr/etc/etc
+
+# 'sim_cpp11_features.pl' backups
+src/groups/mwc/mwcex/*.bak
```
Comment (ddragan-bloomberg) [inline .gitignore:34]: can we generalize that for all packages?

File: src/groups/mwc/mwcex/mwcex_executor.h [hunk 4]
```diff
@@ -295,6 +294,18 @@ class Executor_Box_SboImp {
     /// size of the on-stack buffer used to store the executor target.
     struct Dummy {
         void* d_padding[4];
+
+        bool
+        operator==(BSLS_ANNOTATION_UNUSED const Executor_Box_SboImp::Dummy&)
```
Comment (ddragan-bloomberg) [inline src/groups/mwc/mwcex/mwcex_executor.h:299]: I don't think you'll get warning while using `BSLS_ANNOTATION_UNUSED` if you remove the argument name (like done in this declaration).

File: src/groups/mwc/mwcex/mwcex_executor.h [hunk 5]
```diff
@@ -295,6 +294,18 @@ class Executor_Box_SboImp {
...
+            const BSLS_KEYWORD_NOEXCEPT
+        {
+            return false;
+        }
+
+        void post(BSLS_ANNOTATION_UNUSED const bsl::function<void()>& f) const
```
Comment (ddragan-bloomberg) [inline src/groups/mwc/mwcex/mwcex_executor.h:305]: same as above

File: src/groups/mwc/mwcex/mwcex_executor.h (hunk 4 above)
Comment (ddragan-bloomberg) [reply src/groups/mwc/mwcex/mwcex_executor.h:299]: Also, the argument type can just be `const Dummy&`

**Review by ddragan-bloomberg — APPROVED** [review]
Looks good

**Review by pniedzielski — APPROVED** [review]
